[Push notifications] Fix for IOS 8

DannyvanderJagt opened the pull request:
**Problem**
Using push notifications in IOS 8 will throw this error:
`registerForRemoteNotificationTypes: is not supported in iOS 8.0 and later.`
The problem is that the check is running on compile instead of runtime.

**Solution**
If have changed the compile if statement to a runtime if statement. The fix is tested on: IOS 7.1 and 8.* and everything is working now.
This solution is also discussed in: https://github.com/facebook/react-native/issues/1613 and it was part of  https://github.com/facebook/react-native/pull/1979. (is being separated to keep things moving)

Please let me know what you think.

potsypantsy commented:
:heart:

ehd commented:
Thank you @DannyvanderJagt, this worked for me. Nitpicking: Indentation needs to be adjusted now that it's a statement?

DannyvanderJagt commented:
@ehd Glad it worked and you're right, I have fixed it.

ehd commented:
Thanks again :)

bleonard commented:
This is the fix that i was going to suggest for the issue i'm seeing as well. The issue is that the npm_module seems to define the deploy target itself as 7.0 so the #if will never work - a runtime check like in this PR is necessary.

anthonywebb commented:
Man, really need this fix, no merge yet?

DannyvanderJagt commented:
@anthonywebb No merge yet, but this issue is easy to fix if you don't want to wait. I have also written a tutorial about [How to use push notifications in React-Native for IOS](https://medium.com/@DannyvanderJagt/how-to-use-push-notifications-in-react-native-41e8b14aadae), it also includes an explanation on how to fix this bug yourself.

@nicklockwood @sahrens @brentvatne I know that you are very busy at the moment. However without this fix push notifications aren't working for IOS in React-Native. This fix is also discussed multiple times in the last few months (#1613, #1979, etc) and therefore it should be easy to review and merge this. Let me know if I can help you guys somehow.

jacob-israel-turner commented:
+1, could have used this fix.  :)

magnus-bergman commented:
+1

ericvicenti commented on Libraries/PushNotificationIOS/RCTPushNotificationManager.m:
```diff
@@ -152,18 +152,14 @@ - (void)handleRemoteNotificationsRegistered:(NSNotification *)notification
+    id notificationSettings = [UIUserNotificationSettings settingsForTypes:types categories:nil];
+    [[UIApplication sharedApplication] registerUserNotificationSettings:notificationSettings];
+    [[UIApplication sharedApplication] registerForRemoteNotifications];
+  }else{
```
nit: spaces around else

DannyvanderJagt commented:
@ericvicenti I have added to spaces. However when I did a rebase git added the 'Merge remote-tracking' commit. Let me know if that's a problem.

ide commented:
Can you clean up your branch? Seems like something went wrong when you rebased. Also please squash your commits when you're ready for review.

DannyvanderJagt commented:
@ide I have cleaned up the branch and squashed the commits. Also I have updated the fix, it is now using `app` instead of `[UIApplication sharedApplication]`.

ericvicenti commented:
@facebook-github-bot shipit

jaygarcia commented:
XD. Facebook github bot is stuck on stupid ;)

jibonpab commented on Libraries/PushNotificationIOS/RCTPushNotificationManager.m:
```diff
@@ -157,13 +157,13 @@ - (void)handleRemoteNotificationsRegistered:(NSNotification *)notification
-  [app registerUserNotificationSettings:notificationSettings];
-  [app registerForRemoteNotifications];
-#else
-  [app registerForRemoteNotificationTypes:types];
```
Fifdkl

jibonpab commented on Libraries/PushNotificationIOS/RCTPushNotificationManager.m:
```diff
@@ -157,13 +157,13 @@ - (void)handleRemoteNotificationsRegistered:(NSNotification *)notification
   }
 
   UIApplication *app = RCTSharedApplication();
-#if __IPHONE_OS_VERSION_MIN_REQUIRED >= __IPHONE_8_0
```
Deleted

jibonpab replied:
Dead

vjeux commented:
Whoops, sorry for the spam :x We need to fix that edge case :)

ericvicenti commented:
Oops, that got messy. Landing internally now

(Are you not entertained?!)

ericvicenti commented:
This failed our internal tests due to the following:
<img width="1599" alt="screen shot 2015-09-29 at 5 08 48 pm" src="https://cloud.githubusercontent.com/assets/1483597/10181431/d59e6d90-66cc-11e5-9de5-f85181040e26.png">

@frantic, any thoughts on why the OSS tests pass but the internal ones do not?

DannyvanderJagt commented:
@ericvicenti The OSS is giving a warning instead of an error and therefore it passes. 

<img width="959" alt="screen shot 2015-09-30 at 11 18 24" src="https://cloud.githubusercontent.com/assets/1556953/10189193/11c35996-6765-11e5-90b3-4b4cdf0d7ef2.png">

I have fixed the actual problem/error. The problem is actually that `types` need to be converted to `enum UIUserNotificationType`. To accomplish this we need to replace `types` with  `(NSUInteger)types`.

The result will be: 
```objective-c
  UIApplication *app = RCTSharedApplication();
  if ([app respondsToSelector:@selector(registerUserNotificationSettings:)]) {
    id notificationSettings = [UIUserNotificationSettings settingsForTypes:(NSUInteger)types categories:nil];
    [app registerUserNotificationSettings:notificationSettings];
    [app registerForRemoteNotifications];
  } else {
    [app registerForRemoteNotificationTypes:(NSUInteger)types];
  }
```

I've tested it and it works just as it should.
I would like to hear to hear what you think before I add this.

ericvicenti commented:
Looks great!

Small note: use the `UIUserNotificationSettings` type instead of `id`

DannyvanderJagt commented:
@ericvicenti Thanks for the note.
I have replaced `id notificationSettings` with `UIUserNotificationSettings * notificationSettings`.

I have squashed and rebased the PR.
Let me know if there is anything else I can do.
